### PR TITLE
Allowing transport layer to be customized

### PR DESCRIPTION
**Pull request by abhidnya13**
Separating everything related to requests to http module. Lot of changes around that. I can walk through all the changes.

### CODE REVIEW

File: msal/oauth2cli/default_http_client.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import requests
+
+from .http import HttpClient, Response
+
+
+class DefaultHttpClient(HttpClient):
```
Comment (rayluo) [inline msal/oauth2cli/default_http_client.py]: I understand that there has been multiple efforts on fixing issues in this implementation, like [this initial attempt](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/169/commits/b3a4e09a3cb845cb0244143228a3bc60d405b3fd#diff-102c5027a0791530c6636cd879afe485R46), [this fix](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/169/commits/ccafcf96a15f8d2e9d43dc7d27a3f12cca9125df#diff-102c5027a0791530c6636cd879afe485R44), and the [latest change](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/169/commits/2670e254511a58f6f6b104355a80a21672f4893c#diff-102c5027a0791530c6636cd879afe485R35) (whose interface may or may not be better than the previous one). With all these understanding, let's continue our previous conversation on whether we need to implement and then carry the burden of maintaining such a DefaultHttpClient. Do we really want/need to maintain a DefaultHttpClient, if we do not have to?

In other words, historically msal python uses requests as the underlying http layer, now this task is to "allow customizable http layer (when the customer chooses to do so)". That's it. This task does not imply we have to swap the default http from the battle-tested requests to a home-made http layer.

This DefaultHttpClient can be moved into the test infrastructure, though. It would be useful there.

File: sample/httpx.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        verify=True,  # type: Union[str, True, False, None]
+        proxies=None,  # type: Optional[dict]
+        """
+        self.session = httpx.Client()
+
+    def post(self, url, params=None, data=None, headers=None, **kwargs):
+        if params is None:
+            params = {}
+        params.update(**kwargs)
```
Comment (rayluo) [inline sample/httpx.py]: Is this how "httpx" handle `kwargs`? Have we run e2e test cases with this setup?

File: sample/httpx_client.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import httpx
+from httpx import *
+from msal.oauth2cli import HttpClient, Response
+
+
+class HttpxClient(HttpClient):
```
Comment (rayluo) [inline sample/httpx_client.py]: What is the difference between this implementation and [that one](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/169/files#diff-ec25ae689492ba368f2d6952907acaf4R6)?

File: msal/oauth2cli/oauth2.py [hunk 4]
```diff
@@ -128,7 +123,11 @@ def _obtain_token(  # The verb "obtain" is influenced by OAUTH2 RFC 6749
 
         _data.update(self.default_body)  # It may contain authen parameters
         _data.update(data or {})  # So the content in data param prevails
-        # We don't have to clean up None values here, because requests lib will.
+        filtered = {k: v for k, v in _data.items() if v is not None}
+        _data.clear()
+        _data.update(filtered)
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/oauth2cli/oauth2.py]: I just have it written this was because modifying a dictionary while we are iterating over is not very safe, as mentioned [here](https://stackoverflow.com/questions/6777485/modifying-a-python-dict-while-iterating-over-it/6777632#6777632) and in the official python documentation [here](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects)

File: msal/oauth2cli/oauth2.py [hunk 5]
```diff
@@ -147,23 +146,15 @@ def _obtain_token(  # The verb "obtain" is influenced by OAUTH2 RFC 6749
         if "token_endpoint" not in self.configuration:
             raise ValueError("token_endpoint not found in configuration")
         _headers = {'Accept': 'application/json'}
+        _headers.update(self.default_headers)
         _headers.update(headers or {})
-        resp = (post or self.session.post)(
-            self.configuration["token_endpoint"],
+        resp = (post or self.http_client.post)(self.configuration["token_endpoint"],
             headers=_headers, params=params, data=_data, auth=auth,
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/oauth2cli/oauth2.py]: I think, I would prefer making the necessary changes by comparing the differences and changing things only where needed. For example the dictionary update you mentioned above, I would still prefer how it is implemented here

File: msal/oauth2cli/oauth2.py [hunk 6]
```diff
@@ -147,23 +146,15 @@ def _obtain_token(  # The verb "obtain" is influenced by OAUTH2 RFC 6749
...
-            resp.raise_for_status()  # TODO: Will probably retry here
-        try:
-            # The spec (https://tools.ietf.org/html/rfc6749#section-5.2) says
-            # even an error response will be a valid json structure,
-            # so we simply return it here, without needing to invent an exception.
-            return resp.json()
-        except ValueError:
-            self.logger.exception(
-                    "Token response is not in json format: %s", resp.text)
-            raise
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/oauth2cli/oauth2.py]: I think it got deleted in all these iterations by mistake. Thanks for noticing, I do not intend to remove that. I will restore it to original

File: msal/oauth2cli/default_http_client.py (hunk 1 above)
Comment (abhidnya13) [reply msal/oauth2cli/default_http_client.py]: This implementation still uses the battle-tested requests. It only separates it out to its own layer instead of having it called as and when needed. One of the things which was very confusing when trying to do this task was to actually make sure all the instances of requests are given a functionality of customization. Now you wont touch this layer as it is one of the inner layers similar to oauth2 unless it is needed, but if changes related to requests are needed we will have a single place to make them

File: msal/oauth2cli/oauth2.py [hunk 7]
```diff
@@ -35,13 +34,12 @@ def __init__(
             self,
             server_configuration,  # type: dict
             client_id,  # type: str
+            http_client,  # type: HttpClient
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/oauth2cli/oauth2.py]: According to the current implementation, it is an http_client which can also be a session, session will hold true for session objects only which is what we historically do. But http_client is what we expect, if the passed http client is a session object it still a http_client as we assume it is supposed to have qualities similar to that http_client interface.

File: msal/oauth2cli/oauth2.py [hunk 8]
```diff
@@ -218,10 +209,11 @@ def initiate_device_flow(self, scope=None, timeout=None, **kwargs):
         DAE = "device_authorization_endpoint"
         if not self.configuration.get(DAE):
             raise ValueError("You need to provide device authorization endpoint")
-        flow = self.session.post(self.configuration[DAE],
-            data={"client_id": self.client_id, "scope": self._stringify(scope or [])},
-            timeout=timeout or self.timeout,
-            **kwargs).json()
+        resp = self.http_client.post(self.configuration[DAE],
+                                                   data={"client_id": self.client_id, "scope": self._stringify(scope or [])},
+                                                  timeout=timeout or self.timeout,
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/oauth2cli/oauth2.py]: Sure, I will make most of this absolutely necessary changes

File: msal/oauth2cli/oauth2.py [hunk 9]
```diff
@@ -11,9 +12,7 @@
 import time
 import base64
 import sys
-
-import requests
-
+from .default_http_client import DefaultHttpClient
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/oauth2cli/oauth2.py]: Nice catch, fixing it in the next commit, stay tuned :)

File: sample/httpx.py (hunk 2 above)
Comment (abhidnya13) [reply sample/httpx.py]: Haha, this was supposed to not be checked in but I dint realize pycharm adding it in by default, this is just me trying to use httpx, I'll remove these files in the next commit

File: msal/oauth2cli/oauth2.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rayluo) [reply msal/oauth2cli/oauth2.py]: Thanks for spending your time on research! I totally enjoy such kind of discussion! That's how we grow.

In this topic, there is a subtle difference. [The pattern you found in that Q&A](https://stackoverflow.com/q/6777485) was not safe, because it changes the data being looped, WHILE still inside the loop.

```python
for k,v in d.iteritems():
    del d[f(k)] # remove some item
```

But that is not applicable to our case here. We did NOT modify the data DURING the loop `{k: v for k, v in _data.items() if v is not None}`. We just have an assignment AFTER the loop, and it is just that the new variable is the same old variable, but when the assignment happens, the loop has already finished.

File: msal/oauth2cli/oauth2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rayluo) [reply msal/oauth2cli/oauth2.py]: But, this change is necessary otherwise our implementation would generate exception when customer supplies their http layer which would not have such `auth` behavior (again, because our interface did not require such behavior). We can discuss this in meeting.

[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/oauth2cli/oauth2.py]: Sorry, I dint mean this auth thing, I mean in general with the pulling changes from the oauth branch. I compared these two and  updated the auth behaviour and most of the other things from the existing implementation :)

File: msal/application.py [hunk 10]
```diff
@@ -161,14 +163,18 @@ def __init__(
...
         self.client_credential = client_credential
         self.client_claims = client_claims
-        self.verify = verify
-        self.proxies = proxies
+        if http_client:
+            self.http_client = http_client
+        else:
+            self.http_client = requests.Session()
+            self.http_client.verify = verify
+            self.http_client.proxies = proxies
```
Comment (rayluo) [inline msal/application.py:171]: We mentioned this in a daily sync-up the other day, that the timeout would also have to be handle differently, i.e. solely via http_client. See also [here](https://github.com/AzureAD/microsoft-authentication-library-for-python/compare/oauth2-session?expand=1#diff-ecd2082a45061b914600fa860e2d38bbR84). (From my side, I can go ahead and further change the "session" keyword in that file, so that you could just copy-and-paste it into this PR, and then focus on modifying other relevant things here).

File: tests/test_application.py [hunk 11]
```diff
@@ -50,7 +47,7 @@ class TestClientApplicationAcquireTokenSilentErrorBehaviors(unittest.TestCase):
 
     def setUp(self):
         self.authority_url = "https://login.microsoftonline.com/common"
-        self.authority = msal.authority.Authority(self.authority_url)
+        self.authority = msal.authority.Authority(self.authority_url, http_client= requests.Session())
```
Comment (rayluo) [inline tests/test_application.py:50]: We briefly mentioned this before, that it would actually be valuable to implement a `MinimalHttpClient` (which will be what we had as `DefaultHttpClient` and `HttpResponse`), put it as `tests/http_client.py`, and then we use an instance of such minimal http client to perform all the `test_e2e.py` tests. We will also use a (new) instance of such a `MinimalHttpClient` in `test_authority.py`.

File: msal/mex.py [hunk 12]
```diff
@@ -34,17 +34,16 @@
...
 
-def send_request(mex_endpoint, **kwargs):
-    mex_document = requests.get(
-        mex_endpoint, headers={'Content-Type': 'application/soap+xml'},
-        **kwargs).text
+
+def send_request(mex_endpoint, http_client, **kwargs):
+    resp = http_client.get(mex_endpoint, headers={'Content-Type': 'application/soap+xml'},
+                                    **kwargs)
+    mex_document = resp.text
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/mex.py:46]: https://www.python.org/dev/peps/pep-0008/#blank-lines keeping that extra line, PyCharm prompted according to pep8 suggestion,  will format the rest

File: msal/authority.py [hunk 13]
```diff
@@ -86,18 +85,28 @@ def user_realm_discovery(self, username, correlation_id=None, response=None):
...
-                "https://{netloc}/common/userrealm/{username}?api-version=1.0".format(
-                    netloc=self.instance, username=username),
-                headers={'Accept':'application/json',
-                         'client-request-id': correlation_id},
-                verify=self.verify, proxies=self.proxies, timeout=self.timeout)
-            if resp.status_code != 404:
-                resp.raise_for_status()
-                return resp.json()
-            self.__class__._domains_without_user_realm_discovery.add(self.instance)
-        return {}  # This can guide the caller to fall back normal ROPC flow
```
[the remark this answers is not in the document]
Comment (abhidnya13) [reply msal/authority.py:100]: I'll revert this change as it is not in the scope of this PR. May be we can discuss on why I changed it and get a separate PR later ?

File: msal/oauth2cli/oauth2.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rayluo) [reply msal/oauth2cli/oauth2.py:36]: This entire file looks in good shape. And, as part of the final cross-check, I compare it to my async branch, and notice some line(s) are missing/different. A few of them are actually important.
At this point, rather than pointing them out one-by-one in this PR (which would be very time-consuming) and then you need to hand-pick them, we can probably just download [this file](https://raw.githubusercontent.com/AzureAD/microsoft-authentication-library-for-python/oauth2-session/msal/oauth2cli/oauth2.py) into your laptop, and then, either run a "diff" to see the difference, or you can just use it and see the tests presumably still pass.
(Note that I've changed my previous `session` naming to adopt your `http_client`, so you can use it as a drop-in replacement.)

I also adopt your `http_client` naming to replace the `session` in my another working branch.

File: tests/test_client.py [hunk 14]
```diff
@@ -103,6 +105,7 @@ def setUpClass(cls):
         else:
             cls.client = Client(
                 CONFIG["openid_configuration"], CONFIG['client_id'],
+                http_client=MinimalHttpClient(),
```
Comment (rayluo) [inline tests/test_client.py:108]: Either remove the `http_client=`, or you might as well try to copy [this one](https://raw.githubusercontent.com/AzureAD/microsoft-authentication-library-for-python/oauth2-session/tests/test_client.py). Choose an approach you like better.

File: msal/application.py [hunk 15]
```diff
@@ -139,17 +142,23 @@ def __init__(
...
             Sets the token cache used by this ClientApplication instance.
             By default, an in-memory cache will be created and used.
+        :param http_client: (optional)
+            Your implementation of abstract class HttpClient <msal.oauth2cli.http.http_client>
+            Defaults to a requests session instance
         :param verify: (optional)
             It will be passed to the
             `verify parameter in the underlying requests library
+            This does not apply if you have chosen to pass your own Http client
             <http://docs.python-requests.org/en/v2.9.1/user/advanced/#ssl-cert-verification>`_
```
Comment (rayluo) [inline msal/application.py:152]:
```suggestion
            <http://docs.python-requests.org/en/v2.9.1/user/advanced/#ssl-cert-verification>`_
            This does not apply if you have chosen to pass your own Http client.
```
Nit. Let's swap these two lines. Because, albeit unfortunately unobvious, the original `verify parameter ... <that long url>` part was actually a text with link. Our new sentence should better be NOT part of the link text. Similar suggestion to the `proxies` and `timeout`.
